Added support for starting k brokers instead of 1

asafm opened the pull request:
Also:
- Cleaned up code a bit
- Producer now uses java version of Kafka Producer
- Create topic is shorter

segalziv commented on src/main/java/info/batey/kafka/unit/KafkaUnit.java:
```diff
@@ -15,65 +15,116 @@
 
+    public KafkaUnit withNumOfBrokers(int numOfBrokers) {
+        if (numOfBrokers < 1) {
+            throw new IllegalArgumentException("numOfBrokers must be > 1");
```
numOfBrokers must be >= 1

asafm commented:
@chbatey Any chance you can merge this in ?

Xaerxess commented on src/main/java/info/batey/kafka/unit/KafkaUnit.java:
```diff
@@ -15,65 +15,116 @@
-    private int brokerPort;
-    private Producer<String, String> producer = null;
+    private int brokerPort = -1;
+    private KafkaProducer<String, String> producer = null;
```
Why use `KafkaProducer` here? Not necessary IMO.

asafm replied:
I simply replaced the usage of Producer thus moving to the new API. Are you talking about not needed the KafkaProducer / Producer here all together?

Xaerxess replied:
No, I'm talking about using the interface `Producer` instead of its implementation (`KafkaProducer`).

asafm replied:
Oh no - the package is different as well. The code before used ```kafka.javaapi.producer.Producer``` which was the old producer API (written in Scala). I'm using the new producer API written in Java: ```org.apache.kafka.clients.producer.KafkaProducer```

I never knew KafkaProducer has an interface on top of it but I don't think in such a specific case as this test it matters to use the explicit interface - I don't mind changing it if it is important to you

Xaerxess commented on src/main/java/info/batey/kafka/unit/KafkaUnit.java:
```diff
@@ -98,32 +149,63 @@ private static int parseConnectionString(String connectionString) {
-        zookeeper = new Zookeeper(zkPort);
-        zookeeper.startup();
+        startZookeeper();
+        zkClient = new ZkClient(getZookeeperString(), 30000, 30000, ZKStringSerializer$.MODULE$);
```
What's this `$.MODULE$`?

asafm replied:
Good comment. This took me literally hours to debug why on earth I can't send messages to the brokers. Without it the zkClient simply doesn't write the correct data to ZK. Once you use it, it actually works. I've added the following comment for future developers:
```
        // ZKStringSerializer$.MODULE$ - this obscure object creation is what makes this client work
        // Without it , the topic is reported as created but is actually not usable and produces very hard
        // to understand behaviour
```

Xaerxess commented on src/test/java/info/batey/kafka/unit/MultipleBrokersIntegrationTest.java:
```diff
@@ -0,0 +1,110 @@
+        mapAllTopicPartitionsToSpecificBroker(topic, 2);
+
+        // Wait for reassignment to end
+        Thread.sleep(1000*10);
```
Such long `sleep` in unit tests is not a good practice...

asafm replied:
Good call. I wrote another method which tries every second to see if all parititions replicas have been mapped to desired broker. It does for no more than 10 seconds. Have a look at next commit.

Xaerxess commented on src/main/java/info/batey/kafka/unit/KafkaUnit.java:
```diff
@@ -15,65 +15,116 @@
+        this(1);
     }
 
+    public KafkaUnit(int numOfBrokers) {
```
Having two constructors `KafkaUnit(int)` and `KafkaUnit(int, int)` which do *very* different things is not a good idea. I'd add builder for `KafkaUnit` in such case.

asafm replied:
I marked KafkaUnit(in, int) as @Depracated. I don't think it's really needed. WDYT?

asafm commented:
@chbatey - Any chance you can merge this ?

asafm commented:
@chbatey It's working on my computer. Can you modify the build so it will emit the actual failure? 
```
:testClasses
:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> Process 'Gradle Test Executor 1' finished with non-zero exit value 137

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

Xaerxess commented:
General question: what is use case for multiple brokers? I ask because I feel `kafka-unit` should not replace nor mimic real topology of kafka cluster (that's the case for integration tests, with no `kafka-unit` involved), but rather allow to easily write *unit* tests which communicate with Kafka.

asafm commented:
My code in logz.io failed when we added a broker - partitions were
unavailable since metadata was not refreshed. I fixed it but wanted to have
an integration test for it.
Kafka unit in my view is MiniKafka or StandAlone Kafka which allows to
mimic Kafka in every way a real cluster does. Kafka uses this to test them
selfs. Same goes for Hdfs and base with their MiniHdfs and MiniHbase

On Sun, 10 Jul 2016 at 21:57 Grzegorz Rożniecki <notifications@github.com>
wrote:

> General question: what is use case for multiple brokers? I ask because I
> feel kafka-unit should not replace nor mimic real topology of kafka
> cluster (that's the case for integration tests, with no kafka-unit
> involved), but rather allow to easily write *unit* tests which
> communicate with Kafka.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/chbatey/kafka-unit/pull/29#issuecomment-231604813>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA8Y8eyRjmkPRi8cD21-kvPmdawWnxdjks5qUUCOgaJpZM4JD1Mp>
> .
>

asafm commented:
@Xaerxess Can you help with merging this to master and also fix the build as I wrote above?

asafm commented:
Is someone from the people who have write request to this repo available to merge this pull request?

asafm commented:
We'll go with our version I guess.
